File: livre/src/main/java/fr/deroffal/bibliotheque/livre/adapter/authentification/UserDetailsServiceImpl.java
package fr.deroffal.bibliotheque.livre.adapter.authentification;

import io.vavr.control.Try;
import lombok.RequiredArgsConstructor;
import org.springframework.security.core.userdetails.UserDetails;
import org.springframework.security.core.userdetails.UserDetailsService;
import org.springframework.security.core.userdetails.UsernameNotFoundException;
import org.springframework.stereotype.Service;

@Service
@RequiredArgsConstructor
public class UserDetailsServiceImpl implements UserDetailsService {

    private final UserDetailsClient userDetailsClient;
    private final UserDetailsMapper userDetailsMapper;

    @Override
    public UserDetails loadUserByUsername(final String username) throws UsernameNotFoundException {
        return Try.of(() -> userDetailsClient.loadUserByUsername(username))
            .map(userDetailsMapper::toUserDetails)
            .getOrElseThrow(()-> new UsernameNotFoundException(username));
    }

}

File: livre/src/main/kotlin/fr/deroffal/bibliotheque/livre/adapter/authentification/UserDetailsServiceImpl.kt
package fr.deroffal.bibliotheque.livre.adapter.authentification

import io.vavr.control.Try
import org.springframework.security.core.userdetails.UserDetailsService
import org.springframework.security.core.userdetails.UsernameNotFoundException
import org.springframework.stereotype.Service
import java.util.function.Supplier

@Service
class UserDetailsServiceImpl(
    private val userDetailsClient: UserDetailsClient,
    private val userDetailsMapper: UserDetailsMapper
) : UserDetailsService {

    override fun loadUserByUsername(username: String) =
        Try.of { userDetailsClient.loadUserByUsername(username) }
            .map { userDetailsMapper.toUserDetails(it) }
            .onFailure { println(it) }
            .getOrElseThrow(Supplier { UsernameNotFoundException("Utilisateur $username inconnu !") })

}
